app/models/hint.rb

```diff
@@ -1,5 +1,36 @@
+require 'stringex/core_ext'
+
 class Hint < ApplicationRecord
   def self.match(searchterm)
-    Hint.find_by(fingerprint: searchterm)
+    searchprint = fingerprint(searchterm)
+    Hint.find_by(fingerprint: searchprint)
   end
+
+  # This implements the OpenRefine fingerprinting algorithm. See
+  # https://github.com/OpenRefine/OpenRefine/wiki/Clustering-In-Depth#fingerprint .
+  def self.fingerprint(searchterm)
+    # Preprocess.
+    searchterm.strip!
+    searchterm.downcase!
+    # What's up with the weird regexp? We want to remove punctuation from
+    # the string EXCEPT where it where it is part of the term C++ or C#
+    # (as a standalone term - appearing either immediately after whitespace
+    # or at the beginning of a line). Those terms are meaningfully different
+    # from C, and our users frequently search for them.
+    searchterm.gsub!(/(?<!\s[cC]|\s[cC]\+|^[cC]|^[cC]\+)[[:punct:]]/, '')
```

Comment:
I might move this regex to it's own private method as it's makes reading `self.fingerprint` harder to understand as it's difficult to ignore the regex inline like this. Instead calling `depunctuate` or something would probably make it easier to read and you'd only get caught up in the complicated bits if you actually needed to. Maybe. No big deal either way.